modules/cluster_estimation/cluster_estimation.py

```diff
@@ -0,0 +1,209 @@
+            points_filtered_cluster = np.stack((points_filtered_cluster, clusters[idx]))
+            points_filtered_weights = np.stack((points_filtered_cluster, clusters[idx]))
+            points_filtered_covariances = np.stack((points_filtered_cluster, clusters[idx]))
+        
+        print("Ownership clustering")
+        print(weights)
+        print(covariances)
+        print(clusters)
+        print("-----------------------------------------------")
+
+
+    def run(self, detections: "list[DetectionInWorld]", run_override:  bool) -> "tuple[bool, list[ObjectInWorld | None]]":
+        """
+        Take in list of landing pad detections and return list of estimated landing pad locations
+        if number of detections is sufficient, or if manually forced to run. 
+        """
+
+        # Decide to run
+        if not run_override and not self.decide_to_run(detections):
+            return False, None
+        
+        # Fit points and get cluster data 
+        self.__vgmm_model = self.__vgmm_model.fit(self.__all_points)
+        clusters = self.__vgmm_model.means_
+        covariances = self.__vgmm_model.covariances_
+        weights = self.__vgmm_model.weights_
+
+        # print("Initial:")
+        # print(weights)
+        # print(covariances)
+        # print(clusters)
+        # print("-----------------------------------------------")
+
+        # Sort weights from largest to smallest, along with corresponding covariances and means
+        indices = np.argsort(weights)[::-1]
+        weights = np.sort(weights)[::-1]
+        temp_arr = np.array(covariances)
+        for i in range(len(indices)):
+            covariances[i] = temp_arr[indices[i]]
+        temp_arr = np.array(clusters)
+        for i in range(len(indices)):
+            clusters[i] = temp_arr[indices[i]]
+
+        # print("Sorted:")
+        # print(weights)
+        # print(covariances)
+        # print(clusters)
+        # print("-----------------------------------------------")
+
+        # Loop through each cluster
+        # clusters is a list of centers ordered by weights
+        # most likely cluster listed first in descending weights order
+        objects_in_world = []
+        total_clusters = clusters.shape[0]
+        num_viable_clusters = 1
+        # Drop all clusters after a weight_drop_threshold drop in weight occured
+        while num_viable_clusters < total_clusters:
+            if (weights[num_viable_clusters-1] / weights[num_viable_clusters]) > (1 / self.__weight_drop_threshold):
+                weights = weights[:num_viable_clusters]
+                clusters = clusters[:num_viable_clusters]
+                covariances = covariances[:num_viable_clusters]
+                break
+            num_viable_clusters += 1
+
+        # print("Weight-filtered:")
+        # print(weights)
+        # print(covariances)
+        # print(clusters)
+        # print("-----------------------------------------------")
+
+        # Bad detection removal
```

Comment:
Might be easier to create a new list and add valid clusters rather than remove them in-place.